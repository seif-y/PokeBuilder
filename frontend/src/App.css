.root {
    display: flex;
    min-width: 100%;
}

.main-content {
    flex-grow: 1;
    margin-left: var(--sideNavBarWidth);
    /* Same as width of sidebar*/
}
