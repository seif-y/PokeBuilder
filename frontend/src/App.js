import "./App.css";
import React, { useState, createContext } from "react";
import { BrowserRouter as Router, Switch, Route, Redirect, useParams } from "react-router-dom";
import NOT_FOUND from "./components/global/NOT_FOUND";
import SideNavbar from "./components/nav/SideNavbar";
import Pokedex from "./components/pokedex/Pokedex";
import TeamBuilder from "./components/team-builder/TeamBuilder";
import TeamViewer from "./components/team-viewer/TeamViewer";
import PokemonDetails from "./components/pokedex/pokemon-view/PokemonDetails";
import PokemonDataContextProvider from "./pokeapi/PokemonDataContextProvider";

export const AuthContext = createContext("pokebuilderAuthToken" in localStorage);

function App() {
    const [loggedIn, setLoggedIn] = useState("pokebuilderAuthToken" in localStorage);

    return (
        <Router>
            <AuthContext.Provider value={[loggedIn, setLoggedIn]}>
                <div id="modal-root" className="root">
                    <SideNavbar />
                    <div className="main-content">
                        <PokemonDataContextProvider>
                            <Route path="/pokedex/:name">
                                <GetPokemonDetails />
                            </Route>

                            <Switch>
                                <Route path="/pokedex">
                                    <Pokedex />
                                </Route>
                                <Route path="/team-builder">
                                    <TeamBuilder />
                                </Route>
                                <Route path="/teams">
                                    <TeamViewer />
                                </Route>
                                <Route path="/" exact>
                                    <Redirect to="/pokedex" />
                                </Route>
                                <Route path="*">
                                    <NOT_FOUND />
                                </Route>
                            </Switch>
                        </PokemonDataContextProvider>
                    </div>
                </div>
            </AuthContext.Provider>
        </Router>
    );
}

function GetPokemonDetails() {
    const { name } = useParams();
    return <PokemonDetails name={name}></PokemonDetails>;
}

export default App;
