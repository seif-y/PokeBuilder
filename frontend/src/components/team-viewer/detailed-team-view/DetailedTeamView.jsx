import axios from "axios";
import { useCallback, useContext, useEffect, useState } from "react";
import styles from "./DetailedTeamView.module.css";
import { getToken, getAuthConfig, getUserID } from "../../../util/auth";
import { AuthContext } from "../../../App.js";
import Comment from "./Comment";
import CommentForm from "./CommentForm";
import ImmutableTeamMember from "./ImmutableTeamMember";
import Body from "../util/style-components/Body";
import Headline from "../util/style-components/Headline";
import ShadowedBox from "../util/style-components/ShadowedBox";
import UpvoteBox from "../util/upvotes/UpvoteBox";
import { PokemonDataContext } from "../../../pokeapi/PokemonDataContextProvider";

function Heading({ loggedIn, teamData: { _id: teamID, teamName, creatorUsername, upvotes: initialUpvotes } }) {
    /* TODO duplicate code from TeamViewer */
    /* We highlight upvotes by detecting whether a teamID is included in upvotedTeams */
    const [upvotedTeams, setUpvotedTeams] = useState([]);

    async function fetchAndSetUpvotedTeams(token) {
        const userID = getUserID(token);
        const USER_ENDPOINT = `/api/users/${userID}`;
        const res = await axios.get(USER_ENDPOINT, getAuthConfig(token));
        setUpvotedTeams(res.data.upvotedTeams);
    }

    useEffect(() => {
        if (loggedIn) {
            const token = getToken();
            fetchAndSetUpvotedTeams(token);
        }
    }, [loggedIn]);

    // todo ask backend for a GET upvotes only endpoint?
    const [upvoteCount, setUpvoteCount] = useState(initialUpvotes);

    async function fetchAndSetUpvoteCount() {
        const TEAM_ENDPOINT = `/api/teams/${teamID}`;
        const res = await axios.get(TEAM_ENDPOINT);
        setUpvoteCount(res.data.upvotes);
    }

    async function handleOnVote(isUpvoted) {
        if (loggedIn) {
            const UPVOTE_ENDPOINT = `/api/teams/${teamID}/upvotes`;
            const token = getToken();
            const body = { increment: isUpvoted };
            await axios.patch(UPVOTE_ENDPOINT, body, getAuthConfig(token));
            fetchAndSetUpvotedTeams(token); // renders the highlights
            fetchAndSetUpvoteCount(); // renders the new count
        }
    }

    return (
        <div className={`flex ${styles.heading}`}>
            <UpvoteBox
                isUpvoted={loggedIn && upvotedTeams.includes(teamID)}
                upvotes={upvoteCount}
                onVote={handleOnVote}
            />
            <h1>{`${teamName} by ${creatorUsername}`}</h1>
        </div>
    );
}

function Party({ party = [] }) {
    const allPokemonViews = useContext(PokemonDataContext);
    const [formattedParty, setFormattedParty] = useState([]);

    // TODO: duplicate code from TeamView
    useEffect(() => {
        if (allPokemonViews) {
            setFormattedParty(
                party.map((pokemon) => {
                    const pokemonView = allPokemonViews.find((element) => element.id === pokemon.pokemonID);
                    return Object.assign({}, pokemon, pokemonView);
                })
            );
        }
    }, [party, allPokemonViews]);
    return (
        <div className={styles.sixPack}>
            {formattedParty.map(({ _id, name, notes, sprite, types }) => (
                <ImmutableTeamMember key={_id} name={name} notes={notes} sprite={sprite} types={types} />
            ))}
        </div>
    );
}

function Description({ text }) {
    return (
        <ShadowedBox>
            <Headline>Description</Headline>
            <Body>{text}</Body>
        </ShadowedBox>
    );
}

export default function Render({ teamData }) {
    return teamData ? <DetailedTeamView teamData={teamData} /> : null;
}

function DetailedTeamView({ teamData }) {
    const [loggedIn] = useContext(AuthContext);

    const COMMENTS_ENDPOINT = `/api/teams/${teamData._id}/comments`;
    const [comments, setComments] = useState([]);
    const renderComments = useCallback(async () => {
        const res = await axios.get(COMMENTS_ENDPOINT);
        setComments(res.data);
    }, [COMMENTS_ENDPOINT]);

    useEffect(() => {
        renderComments();
    }, [COMMENTS_ENDPOINT, renderComments]);

    const handleOnSubmitComment = async (comment) => {
        if (comment && loggedIn) {
            const token = getToken();
            await axios.post(COMMENTS_ENDPOINT, { comment }, getAuthConfig(token));
            renderComments();
        }
    };

    return (
        <div className={styles.wrapper}>
            <Heading loggedIn={loggedIn} teamData={teamData} />
            <Party party={teamData.party} />
            <Description text={teamData.description} />
            <CommentForm onSubmitComment={handleOnSubmitComment} />
            <>
                {comments.map(({ _id, comment: body, username }) => (
                    <Comment key={_id} body={body} username={username} />
                ))}
            </>
        </div>
    );
}
