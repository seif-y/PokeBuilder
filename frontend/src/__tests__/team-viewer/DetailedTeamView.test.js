import { AuthContext } from "../../App";
import DetailedTeamView from "../../components/team-viewer/detailed-team-view/DetailedTeamView";
import renderer from "react-test-renderer";

const team = {
    _id: "304u012h5k436h",
    creatorUsername: "TestUser",
    teamName: "Test Team",
    description: "This is not a real team. You have been deceived",
    party: [3, 6, 9, 12, 15, 18].map((id, ind) => {
        return { pokemonID: id, notes: `Team member #${ind}` };
    }),
};

test("DetailedTeamView snapshot for test pokemon matches", () => {
    const tree = renderer
        .create(
            <AuthContext.Provider value={[false, () => {}]}>
                <DetailedTeamView teamData={team} />
            </AuthContext.Provider>
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
