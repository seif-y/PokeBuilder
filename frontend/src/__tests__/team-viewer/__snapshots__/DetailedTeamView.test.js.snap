// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailedTeamView snapshot for test pokemon matches 1`] = `
<div
  className="wrapper"
>
  <div
    className="flex heading"
  >
    <div
      className="flex upvoteBox"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root  arrow clickable MuiSvgIcon-fontSizeLarge"
        focusable="false"
        onClick={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M12 8V4l8 8-8 8v-4H4V8z"
        />
      </svg>
      0
    </div>
    <h1>
      Test Team
    </h1>
    <div
      className="rightHeaderContent"
    >
      <span
        className="username"
      >
        by TestUser
      </span>
    </div>
  </div>
  <div
    className="teamContainer"
  >
    <div
      className="sixPack"
    />
    <div
      className="shadowedBox undefined"
    >
      <div
        className="usernameWrapper"
      >
        <div
          className="header"
          style={
            Object {
              "width": 140,
            }
          }
        >
          Description
        </div>
      </div>
      <div
        className="body undefined"
      >
        This is not a real team. You have been deceived
      </div>
    </div>
  </div>
  <div
    className="commentsWrapper"
  >
    <div
      className="headerWrapper"
    >
      <div
        className="header"
        style={
          Object {
            "paddingLeft": 10,
            "textAlign": "left",
            "width": 400,
          }
        }
      >
        Comments
      </div>
    </div>
    <div
      className="shadowedBox flex commentForm"
    >
      <textarea
        className="textArea form"
        onChange={[Function]}
        onClick={[Function]}
        placeholder="Speak your mind..."
        value=""
      />
      <div
        className="button black"
        onClick={[Function]}
      >
        Submit
      </div>
    </div>
  </div>
</div>
`;
